src/server/worker_load.rs

```diff
@@ -47,75 +85,142 @@ impl WorkerResources {
 // while this structure is also used in hashset for parking resources
 // It is solved in scheduler by directly calling worker.has_time_to_run
 
-#[derive(Debug, Eq, PartialEq, Default)]
+#[derive(Debug, Eq, PartialEq)]
 pub struct WorkerLoad {
     n_cpus: NumOfCpus,
+    n_generic_resources: Vec<GenericResourceAmount>,
 }
 
 impl WorkerLoad {
+    pub fn new(worker_resources: &WorkerResources) -> WorkerLoad {
+        WorkerLoad {
+            n_cpus: 0,
+            n_generic_resources: vec![0; worker_resources.n_generic_resources.len()],
+        }
+    }
+
     #[inline]
     pub fn add_request(&mut self, rq: &ResourceRequest, wr: &WorkerResources) {
         self.n_cpus += wr.n_cpus(rq);
+        for (r, v) in self
+            .n_generic_resources
+            .iter_mut()
+            .zip(wr.n_generic_resources.iter())
+        {
+            *r += v;
+        }
     }
 
     #[inline]
     pub fn remove_request(&mut self, rq: &ResourceRequest, wr: &WorkerResources) {
         self.n_cpus -= wr.n_cpus(rq);
+        for (r, v) in self
+            .n_generic_resources
+            .iter_mut()
+            .zip(wr.n_generic_resources.iter())
+        {
+            *r -= v;
+        }
     }
 
     pub fn is_underloaded(&self, wr: &WorkerResources) -> bool {
         self.n_cpus < wr.n_cpus
+            && self
+                .n_generic_resources
+                .iter()
+                .zip(wr.n_generic_resources.iter())
+                .all(|(v, w)| v < w)
     }
 
     pub fn is_overloaded(&self, wr: &WorkerResources) -> bool {
         self.n_cpus > wr.n_cpus
+            || self
+                .n_generic_resources
+                .iter()
+                .zip(wr.n_generic_resources.iter())
+                .any(|(v, w)| v > w)
     }
 
-    /*pub fn is_overloaded_without_request(
-        &self,
-        wr: &WorkerResources,
-        request: &ResourceRequest,
-    ) -> bool {
-        self.n_cpus > wr.n_cpus
-    }*/
-
     pub fn have_immediate_resources_for_rq(
```

Comment:
This method can only be called if `is_capable_to_run` returned true, right? If so, it should be documented here.

Reply:
It can be called from anywhere (as I can remember)

Reply:
But then `self.n_generic_resources[r.resource as usize]` could panic if the task requests a resource that was introduced after the worker was created?

Reply:
Ah, good catch, it should be get & unwrap

Reply:
unwrap_or(0)